evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs

```diff
@@ -568,6 +612,108 @@ impl<'a, 'b, B: Backend, P: PrecompileSet> CheatcodeStackExecutor<'a, 'b, B, P>
+        runtime: &mut Runtime,
+        address: Address,
+        code: Rc<Vec<u8>>,
+    ) -> ExitReason {
+        let mut forge_runtime = ForgeRuntime::new_with_runtime(runtime, code);
+        let depth = if let Some(depth) = self.state().metadata().depth() { depth + 1 } else { 0 };
+
+        match self.debug_run(&mut forge_runtime, address, depth) {
+            Capture::Exit(s) => s,
+            Capture::Trap(_) => unreachable!("Trap is Infallible"),
+        }
+    }
+
+    pub fn debug_step(&mut self, runtime: &ForgeRuntime, steps: &mut Vec<DebugStep>) -> bool {
+        let pc = if let Ok(pos) = runtime.inner.machine().position() { *pos } else { 0 };
+        let mut push_bytes = None;
+        if let Some((op, stack)) = runtime.inner.machine().inspect() {
+            let wrapped_op = OpCode(op, None);
+            if let Some(push_size) = wrapped_op.push_size() {
+                let push_start = pc + 1;
+                let push_end = pc + 1 + push_size as usize;
+                if push_end < runtime.code.len() {
+                    push_bytes = Some(runtime.code[push_start..push_end].to_vec());
+                } else {
+                    panic!("PUSH{} exceeds codesize?", push_size)
+                }
+            }
+            let mut stack = stack.data().clone();
+            stack.reverse();
+            steps.push(DebugStep {
+                pc,
+                stack,
+                memory: runtime.inner.machine().memory().clone(),
+                op: wrapped_op,
+                push_bytes,
+            });
+            match op {
+                Opcode::CREATE |
+                Opcode::CREATE2 |
+                Opcode::CALL |
+                Opcode::CALLCODE |
+                Opcode::DELEGATECALL |
+                Opcode::STATICCALL => {
+                    // this would create an interrupt, construct a new vec
+                    true
+                }
+                _ => false,
+            }
+        } else {
+            let mut stack = runtime.inner.machine().stack().data().clone();
+            stack.reverse();
+            steps.push(DebugStep {
+                pc,
+                stack,
+                memory: runtime.inner.machine().memory().clone(),
+                op: OpCode(Opcode::INVALID, None),
+                push_bytes,
+            });
+            true
+        }
+    }
+
+    /// Loop stepping the runtime until it stops.
+    fn debug_run(
+        &mut self,
+        runtime: &mut ForgeRuntime,
+        address: Address,
+        depth: usize,
+    ) -> Capture<ExitReason, ()> {
+        let mut done = false;
+        let mut res = Capture::Exit(ExitReason::Succeed(ExitSucceed::Returned));
+        let mut steps = Vec::new();
+        while !done {
+            if self.debug_step(runtime, &mut steps) {
+                self.state_mut().debug_mut().push_node(
+                    0,
+                    DebugNode { address, depth, steps: steps.clone(), ..Default::default() },
```

Comment:
Is this right? Doesn't this push ALL the steps so far, whereas for each Node we'd want only the debug steps corresponding only to the previous call of `debug_step` above?

Reply:
Not sure i totally understand but, since evm execution is recursive, one call's `steps` doesn't exist for a child's call. So there is no intermingling of steps between calls.
i.e.
nodes: [0, 1, 2]
steps 0: A's math prior to call to B == [DebugStep{..}, .., DebugStep{Op::Call()}]
node 1: B's call == [DebugStep{..}, .., DebugStep{Op::Return()}]
node 2: A's handling of the response and extra math == [DebugStep{..}, .., DebugStep{Op::Return()}]

Reply:
of note: we append the new vec of steps to the `debug_mut` *before* having the evm actually execute the operation